lib/gauge_random.cu

```diff
@@ -11,196 +11,183 @@
-	   radius[i]=sqrt( -log(rand2[i]) );
-
-	   temp1[i] = radius[i]*cos(phi[i]);
-	   temp2[i] = radius[i]*sin(phi[i]);
-       }
-
-       ret(0,0) = complex<Float>( temp1[2] + 1./sqrt(3.0)*temp2[3], 0.0);
-       ret(0,1) = complex<Float>( temp1[0], -temp1[1]);
-       ret(0,2) = complex<Float>( temp1[3], -temp2[0]);
-       ret(1,0) = complex<Float>( temp1[0], temp1[1] );
-       ret(1,1) = complex<Float>( -temp1[2] + 1./sqrt(3.0) * temp2[3], 0.0 );
-       ret(1,2) = complex<Float>( temp2[1], -temp2[2] );
-       ret(2,0) = complex<Float>( temp1[3], temp2[0] );
-       ret(2,1) = complex<Float>( temp2[1], temp2[2] );
-       ret(2,2) = complex<Float>( - 2./sqrt(3.0) * temp2[3], 0.0 );
-
-       return ret;
-  }
-
-
-  template<typename Float, typename Gauge>
-  __global__ void computeGenGauss(GaugeGaussArg<Gauge> arg){
-    typedef Matrix<complex<Float>,3> Link;
-    int idx = threadIdx.x + blockIdx.x*blockDim.x;
-    int parity = threadIdx.y + blockIdx.y*blockDim.y;
-
-    typedef Matrix<complex<Float>,3> Link;
-
-
-    if(idx < arg.threads) {
-	int x[4];
-	getCoords(x, idx, arg.X, parity);
-	for (int dr=0; dr<4; ++dr) x[dr] += arg.border[dr]; // extended grid coordinates
+    for (int i = 0; i < 4; ++i) {
+      rand1[i] = Random<real>(localState);
+      rand2[i] = Random<real>(localState);
+    }
 
-	int dx[4] = {0, 0, 0, 0};
-	for(int mu = 0; mu < 4; mu++){
-	    cuRNGState localState = arg.rngstate.State()[idx + parity*arg.threads];
+    for (int i = 0; i < 4; ++i) {
+      phi[i] = 2.0 * M_PI * rand1[i];
+      radius[i] = sqrt(-log(rand2[i]));
+      sincos(phi[i], &temp2[i], &temp1[i]);
+      temp1[i] *= radius[i];
+      temp2[i] *= radius[i];
+    }
 
-	    Link U = genGaussSU3<Float>(localState);
+    // construct Anti-Hermitian matrix
+    ret(0, 0) = complex<real>(0.0, temp1[2] + rsqrt(3.0) * temp2[3]);
+    ret(1, 1) = complex<real>(0.0, -temp1[2] + rsqrt(3.0) * temp2[3]);
+    ret(2, 2) = complex<real>(0.0, -2.0 * rsqrt(3.0) * temp2[3]);
+    ret(0, 1) = complex<real>(temp1[0], temp1[1]);
+    ret(1, 0) = complex<real>(-temp1[0], temp1[1]);
+    ret(0, 2) = complex<real>(temp1[3], temp2[0]);
+    ret(2, 0) = complex<real>(-temp1[3], temp2[0]);
+    ret(1, 2) = complex<real>(temp2[1], temp2[2]);
+    ret(2, 1) = complex<real>(-temp2[1], temp2[2]);
+
+    return ret;
+  }
 
-	    arg.rngstate.State()[ idx + parity*arg.threads ] = localState;
-	    arg.dataDs(mu, linkIndexShift(x,dx,arg.X), parity) = U;
-	}
+  template <typename Float, typename Arg> __global__ void computeGenGauss(Arg arg)
+  {
+    using real = typename mapper<Float>::type;
+    using Link = Matrix<complex<real>, 3>;
+    int x_cb = threadIdx.x + blockIdx.x * blockDim.x;
+    int parity = threadIdx.y + blockIdx.y * blockDim.y;
+    if (x_cb >= arg.threads) return;
+
+    int x[4];
+    getCoords(x, x_cb, arg.X, parity);
+    for (int dr = 0; dr < 4; ++dr) x[dr] += arg.border[dr]; // extended grid coordinates
+
+    if (arg.group && arg.beta == 0.0) {
```

Comment:
I don't have a better idea of how to handle this, but it is sort of weird from a physics standpoint because `beta = 0` is infinitely hot fields :) I guess we're abusing beta a bit, though!

Reply:
Indeed.  We can fix this another day if we have a better idea. 😉

Reply:
I missed that you added `--gaussian-beta` to the command line as the way to expose this, so at least it's consistent (if not maybe a little confusing, but whatever). I guess the better idea would be `sigma`, but it's nothing worth holding this up for :)

Reply:
Just fixed this, and turned beta into sigma.  Took all of 1 minute.